#### FlexiConc
- https://colab.research.google.com/drive/1Y-yXs0RfDWU96Ex-aEYtMCetUtN20Tek?usp=sharing#scrollTo=YGdT4YUb31eK
- https://pypi.org/project/FlexiConc/
- https://clic-fiction.com/flexiconc
- https://github.com/fau-klue/flexiconc-docs
- https://fau-klue.github.io/flexiconc-docs/
- https://github.com/reading-concordances/teaching/tree/main/tutorials/konvens2025
- https://www.dhss.phil.fau.eu/files/2025/03/flexiconc_clic_handout.pdf

In [1]:
#### FlexiConc

In [1]:
# !pip install flexiconc==0.1.5

In [1]:
import flexiconc
dir(flexiconc)

['CONFIG',
 'Concordance',
 'ConfigParser',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 'builtins',
 'concordance',
 'jsonschema_validate',
 'load_config',
 'os',
 'subprocess',
 'user_config_dir']

In [2]:
import flexiconc
print(flexiconc.__file__)
print(flexiconc.__name__)

/Users/vivien/opt/anaconda3/envs/pytorch/lib/python3.13/site-packages/flexiconc/__init__.py
flexiconc


In [3]:
import importlib.metadata

print(importlib.metadata.version("flexiconc"))
print(importlib.metadata.metadata("flexiconc"))

0.1.5
Metadata-Version: 2.4
Name: FlexiConc
Version: 0.1.5
Summary: A Python package to assist concordance reading
Home-page: https://github.com/fau-klue/flexiconc
Author: =?utf-8?q?RC21_=28Nathan_Dykes=2C_Stephanie_Evert=2C_Michaela_Mahlberg=2C_Alexander_Piperski_=40_Friedrich-Alexander-Universit=C3=A4t_Erlangen-N=C3=BCrnberg=29?=
Author-email: aleksandr.piperski@fau.de
Classifier: Programming Language :: Python :: 3
Classifier: License :: OSI Approved :: MIT License
Classifier: Operating System :: OS Independent
Requires-Python: >=3.6
Description-Content-Type: text/markdown
Requires-Dist: anytree>=2.12.1
Requires-Dist: networkx>=3.1
Requires-Dist: numpy<3.0,>=2.0
Requires-Dist: pandas<3.0,>=2.2.2
Requires-Dist: platformdirs>=4.3
Requires-Dist: jsonschema>=4.23.0
Dynamic: author
Dynamic: author-email
Dynamic: classifier
Dynamic: description
Dynamic: description-content-type
Dynamic: home-page
Dynamic: requires-dist
Dynamic: requires-python
Dynamic: summary
Description: =?utf-8?q?=23_I

#### Notebook RC21

In [4]:
from flexiconc import Concordance
c = Concordance()

In [5]:
c.retrieve_from_clic(api_base_url="https://clic-fiction.com/api/concordance", query=["stomach"], corpora="dickens")

In [6]:
c.root

AnalysisTreeNode(bookmarked=False, concordance=<weakref at 0x12c63ccc0; to 'flexiconc.concordance.Concordance' at 0x12c5556a0>, id=0, label='root', line_count=51, node_type='subset', selected_lines=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50])

In [7]:
from IPython.display import HTML, display
from flexiconc.visualization.html_visualizer import generate_concordance_html, generate_analysis_tree_html
display(HTML(generate_analysis_tree_html(c)))

In [8]:
display(HTML(generate_concordance_html(c, c.root, n=1000)))

In [9]:
algos = c.root.available_algorithms()
sorted(list(algos.keys()))

['Flat Clustering by Embeddings',
 'KWIC Grouper Ranker',
 'Manual Line Selection',
 'Partition by Metadata Attribute',
 'Partition by Ngrams',
 'Partition with OpenAI',
 'Random Sample',
 'Random Sort',
 'Select Slot',
 'Select Weighted Sample by Metadata',
 'Select by Metadata Attribute',
 'Select by Rank',
 'Select by Sort Keys',
 'Select by a Token-Level Attribute',
 'Set Operation',
 'Sort by Corpus Position',
 'Sort by Token-Level Attribute']

In [10]:
n1 = c.root.add_subset_node(
    ("Random Sample",
     {'sample_size': 10, 'seed': 9101989})
)
display(HTML(generate_concordance_html(c, n1, n=1000)))

In [11]:
display(HTML(generate_analysis_tree_html(c)))

In [12]:
c.root.schema_for("Random Sample")

{'type': 'object',
 'properties': {'sample_size': {'type': 'integer',
   'description': 'The number of lines to sample.',
   'minimum': 1,
   'x-eval': 'dict(maximum=node.line_count)',
   'maximum': 51},
  'seed': {'type': ['integer'],
   'description': 'An optional seed for random number generation.',
   'default': 42}},
 'required': ['sample_size', 'seed']}

In [13]:
display(HTML(generate_concordance_html(c, n1, n=1000, metadata_columns=["text_id"])))

In [14]:
c.root.schema_for('Select by Metadata Attribute')

{'type': 'object',
 'properties': {'metadata_attribute': {'type': 'string',
   'description': 'The metadata attribute to filter on.'},
  'value': {'type': ['string', 'number', 'array'],
   'description': 'The value to compare against, or a list of acceptable values. When a list is provided, only equality is used.'},
  'operator': {'type': 'string',
   'enum': ['==', '<', '<=', '>', '>='],
   'description': "The comparison operator for numeric comparisons. Only allowed for single numeric values. Default is '=='.",
   'default': '=='},
  'regex': {'type': 'boolean',
   'description': 'If True, use regex matching for string comparisons (only with equality). Default is False.',
   'default': False},
  'case_sensitive': {'type': 'boolean',
   'description': 'If True, perform case-sensitive matching for strings. Default is False.',
   'default': False},
  'negative': {'type': 'boolean',
   'description': 'If True, invert the selection. Default is False.',
   'default': False}},
 'required': 

In [15]:
n2 = c.root.add_subset_node(
    ("Select by Metadata Attribute",
     {'metadata_attribute': "text_id", 'value': "OMF"})
)
display(HTML(generate_concordance_html(c, n2, n=1000, metadata_columns=["text_id"])))

In [16]:
display(HTML(generate_analysis_tree_html(c)))

In [17]:
n2a = n2.add_arrangement_node(
    ordering = [
        ("Sort by Token-Level Attribute",
         {'tokens_attribute': "word", 'sorting_scope': "token", 'offset': -1})
    ]
)
display(HTML(generate_analysis_tree_html(c)))
display(HTML(generate_concordance_html(c, n2a, n=1000, metadata_columns=["text_id"])))

In [18]:
n3 = c.root.add_arrangement_node(
    grouping = ("Partition by Ngrams",
         {'positions': [-2, -1]})
)
display(HTML(generate_analysis_tree_html(c)))
display(HTML(generate_concordance_html(c, n3, n=1000, metadata_columns=["text_id"])))

In [19]:
n4 = n3.add_subset_node(
    ("Manual Line Selection",
     {'groups': ["('a', 'proud')"]})
)
display(HTML(generate_analysis_tree_html(c)))
display(HTML(generate_concordance_html(c, n4, n=1000, metadata_columns=["text_id"])))

In [20]:
c.add_annotation(
    ("Annotate with Sentence Transformers",
     {'window_start': -10, 'window_end': 10, 'model_name': 'all-MiniLM-L6-v2'})
)

In [21]:
n5 = c.root.add_arrangement_node(
    grouping = ("Flat Clustering by Embeddings",
                {"embeddings_column": "embeddings_sentence_transformers", "n_partitions": 5}),
    ordering = [
        ("Sort by Token-Level Attribute", {'offset': -1}),
        ("Sort by Token-Level Attribute", {'sorting_scope': "right"})
    ]
)
display(HTML(generate_concordance_html(c, n5, n=1000, metadata_columns=["text_id"])))

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


In [22]:
display(HTML(generate_analysis_tree_html(c)))

Jetzt kommt der eigene Versuch:

In [1]:
from flexiconc import Concordance
concordance = Concordance()

In [5]:
concordance.retrieve_from_cwb( 
    registry_dir="/usr/local/share/cwb/registry",
    corpus_name="TRUMP",
    query='[word="great"]',
    tokens_attrs=["word", "lemma", "pos", "x_pos", "lemma_p"],
    metadata_attrs=["text_id", "text_date", "text_time", "text_day", "text_month", "text_year", "text_favorites", "text_retweets", "text_isRetweet", "text_isDeleted", "text_device", "text_isFlagged", "text_platform"],
    context=20
)

/Users/vivien/opt/anaconda3/envs/flexiconc/lib/python3.12/site-packages/flexiconc/utils/retrieve.py:97: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  matches = tokens[tokens['offset'] == 0].groupby('line_id').apply(


In [ ]:
C = Concordance() C.load_from_cqpweb_export("CQPweb_WaterSanitation_ParlUK.txt")

In [ ]:
trump = Corpus("TRUMP", backend=backend)
musk = Corpus("MUSK", backend=backend)

In [ ]:
results = musk.concordance("Tesla")
print(results.head())

In [ ]:
results = trump.concordance('[pos="NN"]')
print(results.head())

| `xpos` (Tweebank / PTB) | `upos` (Universal) | Bedeutung / Beispiele                                                                  |
| ----------------------- | ------------------ | -------------------------------------------------------------------------------------- |
| **NN**                  | NOUN               | Noun, singular → *dog, idea*                                                           |
| **NNS**                 | NOUN               | Noun, plural → *dogs, cars*                                                            |
| **NNP**                 | PROPN              | Proper noun, singular → *Trump, Canada*                                                |
| **NNPS**                | PROPN              | Proper noun, plural → *the Smiths*                                                     |
| **PRP**                 | PRON               | Personal pronoun → *I, you, he*                                                        |
| **PRP\$**               | PRON               | Possessive pronoun → *my, your*                                                        |
| **WP**                  | PRON               | Wh-pronoun → *who, what*                                                               |
| **WP\$**                | PRON               | Possessive wh-pronoun → *whose*                                                        |
| **DT**                  | DET                | Determiner → *the, a, some*                                                            |
| **PDT**                 | DET                | Predeterminer → *all the kids*                                                         |
| **WDT**                 | DET                | Wh-determiner → *which*                                                                |
| **JJ**                  | ADJ                | Adjective → *big, nice*                                                                |
| **JJR**                 | ADJ                | Comparative adj → *bigger*                                                             |
| **JJS**                 | ADJ                | Superlative adj → *biggest*                                                            |
| **RB**                  | ADV                | Adverb → *quickly*                                                                     |
| **RBR**                 | ADV                | Comparative adv → *faster*                                                             |
| **RBS**                 | ADV                | Superlative adv → *fastest*                                                            |
| **WRB**                 | ADV                | Wh-adverb → *how, when, why*                                                           |
| **VB**                  | VERB               | Verb base → *eat, go*                                                                  |
| **VBD**                 | VERB               | Verb past → *ate, went*                                                                |
| **VBG**                 | VERB               | Verb gerund/participle → *eating*                                                      |
| **VBN**                 | VERB               | Verb past participle → *eaten*                                                         |
| **VBP**                 | VERB               | Verb non-3sg present → *eat, go*                                                       |
| **VBZ**                 | VERB               | Verb 3sg present → *eats, goes*                                                        |
| **MD**                  | AUX                | Modal → *can, should*                                                                  |
| **IN**                  | ADP                | Preposition, subordinating → *in, of, because*                                         |
| **TO**                  | PART               | Particle *to*                                                                          |
| **CC**                  | CCONJ              | Coordinating conj → *and, or*                                                          |
| **UH**                  | INTJ               | Interjection → *oh, hi*                                                                |
| **EX**                  | PRON               | Existential *there*                                                                    |
| **FW**                  | X                  | Foreign word                                                                           |
| **SYM**                 | SYM                | Symbol → *%, \$, +*                                                                    |
| **LS**                  | X                  | List item marker                                                                       |
| **CD**                  | NUM                | Cardinal number → *5, twenty*                                                          |
| **POS**                 | PART               | Possessive marker *’s*                                                                 |
| **RP**                  | PART               | Particle → *up, off*                                                                   |
| **ADD**                 | PROPN              | **Spezialtag Tweebank**: URL, Email, @mention, Hashtag → *@user, #hashtag, https\://…* |
| **NFP**                 | SYM                | **Spezialtag Tweebank**: Non-functional punctuation → *🤣, :), ❤️*                     |
| **. , : ; - etc.**      | PUNCT              | Satzzeichen                                                                            |


In [ ]:
# Löschen der Kernel: 
# jupyter kernelspec uninstall tpann

In [ ]:
 from flexiconc import TextImport T = TextImport()
T.load_files(
paths=["data"], shorten_paths=True, db_name="db.sqlite", use_spacy=True, lemma=True, pos=True,
tag=True )
C = T.concordance_from_query("stomach")

4.3 Einordnung und Analyse Trump
Historisch, Politischer Einblick
4 Seiten
- Einordnung der Aktivität auf TruthSocial & Twitter
- Meinungswechsel (Selensky, Putin) Ukrainekrieg, Gaza
- Social Media Stil, Persönlichkeitsstruktur, Personenkult, MAGA
- Allgemeine Auffälligkeiten: sehr starke Kontextabhängigkeit (wie vielleicht oft bei Tweets?)
- Wahl 2016 & 2024
- Verwendung von Vornamen & Nachnamen, Schwangerschaft, Frauen
- Russland, Ukraine, andere Kriege
- Trumpish, fake News, Truth
- „low IQ“ -> Abwertung der Intelligenz, „stupid“, „lightweight“
- Plattform für Verschwörungen: „Deep State“
4.4 Einordnung und Analyse Musk
Größtenteils relevant ab 2022, Kauf von Twitter; ab da auch viel mehr gepostet
3 Seiten
- Meinungsfreiheit
- Oligarch
- Geld und Plutokratie
4.5 Korpusanalyse Vergleich Trump - Musk
Streit Musk, Trump (vorher vs. nachher), „woke“, 2 Seiten
